File: src/routes/wellKnown.js
const express = require('express');
const router = express.Router();
const cors = require('cors');
const { getJwks } = require('../crypto/keys');
const oauthService = require('../services/oauthService');

// All .well-known discovery and key endpoints must allow unrestricted cross-origin requests
// so LLM clients and MCP servers can discover and verify anywhere.
router.use(cors({ origin: '*' }));

/**
 * RFC 8414 OAuth 2.0 / 2.1 Authorization Server Metadata
 */
router.get('/oauth-authorization-server', (req, res) => {
  res.setHeader('Content-Type', 'application/json');
  res.setHeader('Cache-Control', 'public, max-age=3600');
  res.json(oauthService.getAuthorizationServerMetadata());
});

/**
 * OAuth 2.1 Protected Resource Metadata (PRM)
 * draft-ietf-oauth-resource-metadata
 */
router.get('/oauth-protected-resource', (req, res) => {
  const resource = req.query.resource;
  res.setHeader('Content-Type', 'application/json');
  res.setHeader('Cache-Control', 'public, max-age=3600');
  res.json(oauthService.getProtectedResourceMetadata(resource));
});

/**
 * RFC 7517 JSON Web Key Set (JWKS)
 * Contains the public RS256 key(s) used by MCP servers to verify signatures
 */
router.get('/jwks.json', (req, res) => {
  res.setHeader('Content-Type', 'application/json');
  res.setHeader('Cache-Control', 'public, max-age=300'); // 5 minutes cache
  res.json(getJwks());
});

module.exports = router;

File: src_py/routers/well_known.py
from fastapi import APIRouter, Response, Query
from typing import Optional
from src_py.crypto.keys import get_jwks
from src_py.services import oauth_service

router = APIRouter(prefix="/.well-known", tags=["Discovery & Keys"])


@router.get("/oauth-authorization-server")
def get_oauth_authorization_server(response: Response):
    """RFC 8414 OAuth 2.0 / 2.1 Authorization Server Metadata."""
    response.headers["Cache-Control"] = "public, max-age=3600"
    return oauth_service.get_authorization_server_metadata()


@router.get("/oauth-protected-resource")
def get_oauth_protected_resource(response: Response, resource: Optional[str] = Query(None)):
    """OAuth 2.1 Protected Resource Metadata (PRM)."""
    response.headers["Cache-Control"] = "public, max-age=3600"
    return oauth_service.get_protected_resource_metadata(resource)


@router.get("/jwks.json")
def get_jwks_json(response: Response):
    """RFC 7517 JSON Web Key Set (JWKS) containing the public RS256 key(s)."""
    response.headers["Cache-Control"] = "public, max-age=300"
    return get_jwks()
